teps/0088-result-summaries.md

```diff
@@ -0,0 +1,392 @@
+- [Implementation Pull request(s)](#implementation-pull-request-s)
+- [References (optional)](#references-optional)
+  <!-- /toc -->
+
+## Summary
+
+<!--
+This section is incredibly important for producing high quality user-focused
+documentation such as release notes or a development roadmap.  It should be
+possible to collect this information before implementation begins in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Currently, Tekton Results are modelled so that Records keep a reference to their
+parent Results, but Results themselves do not have a reference back to their
+child Records.
+
+```
++--------+               +--------+
+|        |               |        |
+| Result | <-----+------ | Record |
+|        |       |       |        |
++--------+       |       +--------+
+                 |
+                 |       +--------+
+                 |       |        |
+                 +------ | Record |
+                         |        |
+                         +--------+
+```
+
+This model lets us store and update arbitrary numbers of Records without causing
+contention on the shared parent Result. However, this model has a disadvantage -
+it is difficult to determine which Record is the primary / root Record for a
+given Result without scanning through all Records. This makes it harder to
+determine things like overall Result status, timings, and other information that
+might be common to all Results.
+
+To address this, we wish to add a "Record Summary" to a Result to allow for
+referencing and distilling common fields into a Result without needing to search
+through Records.
+
+```proto
+message Result {
+  ...
+  RecordSummary record_summary = 5;
+}
+
+message RecordSummary {
+  // The name of the Record this summary represents.
+  string record = 1  [
+   (google.api.resource_reference).type = "results.tekton.dev/Record"
+  ];
+  // Identifier of underlying data.
+  // e.g. `pipelines.tekton.dev/PipelineRun`
+  string type = 2;
+
+  // Common Record agnostic fields.
+  google.protobuf.Timestamp start_time = 3;
+  google.protobuf.Timestamp end_time = 4;
+
+  enum Status {
+    UNKNOWN = 1;
+    SUCCESS = 2;
+    FAILURE = 3;
+    ...
+  }
+  Status status = 5;
+
+   // Key-value pairs representing arbitrary underlying record data that clients want to include
+   // that aren't covered by the above Record agnostic fields.
+   map<string, string> record_data = 6;
```

Comment:
Couple questions on this field:
- Is there a specific use-case in mind for it? Given the non-goal of storing detailed record information I wonder if this is needed.
- Who / how does data get recorded into this map? Is it configurable in some way or does the operator need to write code to record to it?  Or something else?